plugins/vli/fu-vli-pd-parade-device.c

```diff
@@ -0,0 +1,668 @@
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_read (self, self->page7_addr, 0x00, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* format as a hex triplet */
+	version_str = g_strdup_printf ("%02X.%02X.%02X", buf[0], buf[1], buf[2]);
+	fu_device_set_version (FU_DEVICE (self), version_str, FWUPD_VERSION_FORMAT_TRIPLET);
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_set_wp (FuVliPdParadeDevice *self, gboolean val, GError **error)
+{
+	return fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0xB3,
+						  val ? 0x10 : 0x00, error);
+}
+
+static gboolean
+fu_vli_pd_parade_device_write_enable (FuVliPdParadeDevice *self, GError **error)
+{
+	/* Set_WP_High, SPI_WEN_06, Len_00, Trigger_Write, Set_WP_Low */
+	if (!fu_vli_pd_parade_device_set_wp (self, TRUE, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x90, 0x06, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x92, 0x00, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x93, 0x05, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_set_wp (self, FALSE, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_write_disable (FuVliPdParadeDevice *self, GError **error)
+{
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0xDA, 0x00, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_write_status (FuVliPdParadeDevice *self, guint8 target_status, GError **error)
+{
+	/* Set_WP_High, SPI_WSTS_01, Target_Status, Len_01, Trigger_Write, Set_WP_Low */
+	if (!fu_vli_pd_parade_device_set_wp (self, TRUE, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x90, 0x01, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x90, target_status, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x92, 0x01, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x93, 0x05, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_set_wp (self, FALSE, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_wait_ready (FuVliPdParadeDevice *self, GError **error)
+{
+	gboolean ret = FALSE;
+	guint limit = 100;
+	guint8 buf = 0x0;
+
+	/* Wait_For_SPI_ROM_Until_Not_Busy */
+	for (guint wait_cnt1 = 0; wait_cnt1 < limit; wait_cnt1++) {
+		buf = 0xFF;
+		if (!fu_vli_pd_parade_device_i2c_read (self, self->page_addr,
+						       0x9E, &buf, sizeof(buf),
+						       error))
+			return FALSE;
+		/* Busy_Status, bit[1,0]:Byte_Program, bit[3,2]:Sector_Erase, bit[5,4]:Chip_Erase */
+		if ((buf & 0x0C) == 0) {
+			ret = TRUE;
+			break;
+		}
+	}
+	if (!ret) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "Wait_9e_TOut");
```

Comment:
Not really a user friendly string if this situation happens

Reply:
Fixed.